Dockerfiles release

brycem opened the pull request:
Adding Windows Dockerfiles for stable releases

brycem committed:
Merge remote-tracking branch 'refs/remotes/PowerShell/master'

brycem committed:
Release versions of WindowsServerCore and Nano DockerFiles

Embedded POWERSHELL_* ARGs will need to be updated when new releases
become available

msftclas commented:
Hi __@brycem__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <p>
        It looks like you're a Microsoft contributor (Bryce Milton). If you're full-time, we DON'T require a Contribution License Agreement. If you are a vendor, please DO sign the electronic Contribution License Agreement. It will take 2 minutes and there's no faxing! https://cla.microsoft.com.
    </p>

TTYL, MSBOT;

brycem commented:
Resubmitting from personal fork as requested in https://github.com/PowerShell/PowerShell/pull/2461

brycem commented on docker/release/nanoserver/Dockerfile:
```diff
@@ -0,0 +1,59 @@
+LABEL Readme.md="https://github.com/PowerShell/PowerShell/blob/master/docker/README.md"
+LABEL Description="This Dockerfile will install and build the latest reslease of PS on NanoServer."
+
+ARG POWERSHELL_ZIP=https://github.com/PowerShell/PowerShell/releases/download/v6.0.0-alpha.10/powershell-6.0.0-alpha.10-win10-x64.zip
```
These two ARGs would need to be edited when the next release comes out.  Do you need a Pester test that'd break if it doesn't match?

brycem commented on docker/release/nanoserver/Dockerfile:
```diff
@@ -0,0 +1,59 @@
+# escape=`
```
This change the Docker escape char to match Powershell's so that it makes it less tedious to in-line powershell code in a Dockerfile.

vors replied:
As far as I see, that causes to create a single line for the whole PS code. It's a little bit tedious to have a end-of-line escaping on every line. Even in vanilla powershell it's common to prefer **splatting** over end-of-line escaping, because it's more  maintainable. What happens, if you just remove the backticks from the whole file?

brycem replied:
@vors - I double-checked my thinking on this with @raghushantha this afternoon and he agreed that using this docker processor directive aids in making the dockerfile simpler to read\maintain.  If you look at the way Docker is executing powershell, it's doing it with a -command param, so all that follows is processed as a single lines with standard Powershell escapes.

vors replied:
Oh, I see now. Would move this logic to `bootstrap.ps1` file (that's how it's done in [existing](https://github.com/PowerShell/PowerShell/blob/92b0eb336d7af74d5e439f58ffd3a367d511c0e5/docker/nightly/centos7/Dockerfile#L8) Dockerfiles) make sense?

brycem replied:
I refactored the two docker files into docker+bootstrap files here:  https://github.com/brycem/PowerShell/tree/docker/docker/release - They work fine off a local share, but because of the way Docker build context works, you lose the user simplicity of doing 'docker build https://raw.githubusercontent.com/brycem/PowerShell/docker/docker/release/windowsservercore/Dockerfile' like you could with the current self-contained dockerfile format...  I'd recommend refactoring the Linux files to inline the code for the more flexible usability.

brycem replied:
@andschwa  I've created in-line script versions of the Linux nightly builds here:  https://github.com/brycem/PowerShell/tree/docker/docker/nightly  Example:  'docker build https://raw.githubusercontent.com/brycem/PowerShell/docker/docker/nightly/centos7/Dockerfile'

andschwa replied:
Yeah the inline makes using it simpler, it's just more duplicate code. Granted I ended up having to copy the bootstrap script into each folder anyway... I'm fine with changing it to be inline.

brycem replied:
Ok - had to add a SHELL ["powershell", "-command"] line, but otherwise they seem to work.  Testing, then I'll resubmit with the in-lined Linux docker files

brycem reviewed (commented):
On a RS\Win10 machine with a recent build of the Docker engine files, you can build this with the following command:  docker build --isolation hyperv --rm https://raw.githubusercontent.com/brycem/PowerShell/3ed4c397b5b395355f4284b2dcc3531b31cfec2f/docker/release/nanoserver/Dockerfile

brycem committed:
Updating to use Alpha-11 URLs\SHA

Updated URLs to point at Alpha 11 URLs, reduced number of intermediate
stages, simplified logging code, and added a new %PSCORE% environment
variable to make it easier to launch from other shells

lzybkr commented on docker/release/nanoserver/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+FROM microsoft/nanoserver
+MAINTAINER brycem@microsoft.com
+LABEL Readme.md="https://github.com/PowerShell/PowerShell/blob/master/docker/README.md"
+LABEL Description="This Dockerfile will install and build the latest reslease of PS on NanoServer."
```
s/reslease/release/

brycem committed:
Updating to Alpha 12 and correcting 'reslease' typo

brycem committed:
Refactoring Linux Nighly Dockerfiles to in-line Bootstrap.ps1 code

Provides simpler usability by making the dockerfiles self-contained.
